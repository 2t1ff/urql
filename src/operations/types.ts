import { DocumentNode, OperationDefinitionNode } from 'graphql';
import { FragmentSelectionSets, VarsMap } from '../ast';
import { Store } from '../store';
import { Scalar, SystemFields } from '../types';

export interface DataFields {
  [fieldName: string]: Data | Data[] | Scalar;
}

export type Data = DataFields & SystemFields;

export interface Request {
  query: DocumentNode;
  variables?: null | object;
}

export interface Result {
  data?: null | Data;
  isComplete: boolean;
  dependencies: string[];
}

export interface Context {
  /** List of heuristic keys indicating what paths have been walked */
  dependencies: string[];
  /** When reading from cache it indicates whether the result is likely complete */
  isComplete: boolean;
  operation: OperationDefinitionNode;
  store: Store;
  fragments: FragmentSelectionSets;
  vars: VarsMap;
}
