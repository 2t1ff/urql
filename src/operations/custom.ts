import { FieldResolver } from '../types';

export const makeCustomResolver = (
  innerResolver: FieldResolver
): FieldResolver => {
  return (fieldName, rootValue, args, context, info) => {
    const { __typename: typeName } = rootValue;
    const fieldResolver = context.store.getResolver(typeName, fieldName);
    if (fieldResolver !== undefined) {
      const result = fieldResolver(args, context, info);
      if (result !== undefined) {
        return result;
      }
    }

    return innerResolver(fieldName, rootValue, args, context, info);
  };
};
