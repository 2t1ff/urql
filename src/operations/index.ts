export { query } from './query';
export { write } from './write';
export { gc } from './gc';
