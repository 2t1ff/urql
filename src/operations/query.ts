import graphql from '../exec';
import { keyForLink, keyOfEntity } from '../keys';
import Store from '../store';
import { FieldResolver, Link, Request, Result } from '../types';

const entityOfLink = (store: Store, link: Link) => {
  if (Array.isArray(link)) {
    // @ts-ignore
    return link.map(inner => entityOfLink(store, inner));
  } else if (link === null || typeof link !== 'string') {
    return null;
  }

  return store.getEntity(link);
};

const queryResolver: FieldResolver = (
  fieldName,
  rootValue,
  args,
  context,
  info
) => {
  if (info.isLeaf) {
    return rootValue[fieldName];
  }

  const parentKey = keyOfEntity(rootValue);
  if (parentKey === null) {
    context.isComplete = false;
    return null;
  }

  const { store } = context;
  const link = store.getLink(keyForLink(parentKey, fieldName, args));
  if (link === null || link === undefined) {
    const fieldValue = rootValue[fieldName];
    if (fieldValue === undefined) {
      context.isComplete = false;
      return null;
    }

    return fieldValue;
  }

  const entity = entityOfLink(store, link);
  if (entity === null) {
    context.isComplete = false;
  }

  return entity;
};

const query = (store: Store, request: Request): Result => {
  const context = { isComplete: true, store };
  const response = graphql(
    queryResolver,
    request,
    store.getOrCreateEntity('Query'),
    context
  );

  return {
    dependencies: store.flushTouched(),
    isComplete: context.isComplete,
    response,
  };
};

export default query;
