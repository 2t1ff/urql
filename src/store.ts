import { Entity, EntityMap, KeyExtractor, Link, LinkMap } from './types';
import {
  assignObjectToMap,
  isOperation,
  keyOfEntity,
  objectOfMap,
} from './utils';

export interface StoreData {
  records: {
    [key: string]: Entity;
  };
  links: {
    [key: string]: Link;
  };
}

export interface StoreOpts {
  initial?: StoreData;
  keyExtractor?: KeyExtractor;
}

class Store {
  private touched: string[];
  private records: EntityMap;
  private links: LinkMap;

  public keyOfEntity: (entity: Entity) => null | string;

  constructor({ initial, keyExtractor }: StoreOpts) {
    this.touched = [];
    this.records = new Map();
    this.links = new Map();

    if (keyExtractor !== undefined) {
      this.keyOfEntity = (entity: Entity) => {
        const key = keyExtractor(entity);
        return key !== undefined ? key : keyOfEntity(entity);
      };
    } else {
      this.keyOfEntity = keyOfEntity;
    }

    if (initial !== undefined) {
      assignObjectToMap(this.records, initial.records);
      assignObjectToMap(this.links, initial.links);
    }
  }

  getEntity(key: string): null | Entity {
    if (!isOperation(key)) {
      this.touched.push(key);
    }

    const entity = this.records.get(key);
    return entity !== undefined ? entity : null;
  }

  getOrCreateEntity(key: string): Entity {
    const prev = this.records.get(key);
    if (prev !== undefined && prev !== null) {
      return prev;
    }

    const entity = Object.create(null);
    this.records.set(key, entity);
    return entity;
  }

  writeEntityValue(key: string, prop: string, val: any) {
    if (!isOperation(key)) {
      this.touched.push(key);
    }

    const entity = this.getOrCreateEntity(key);
    if (val === undefined) {
      delete entity[prop];
    } else {
      entity[prop] = val;
    }
  }

  getLink(key: string): Link {
    this.touched.push(key);
    const link = this.links.get(key);
    return link !== undefined ? link : null;
  }

  writeLink(key: string, link: Link) {
    this.touched.push(key);

    if (link === null) {
      this.links.delete(key);
    } else {
      this.links.set(key, link);
    }
  }

  toJSON(): StoreData {
    return {
      records: objectOfMap(this.records),
      links: objectOfMap(this.links),
    };
  }

  flushTouched(): string[] {
    const touched = this.touched.filter((key, i, arr) => {
      return arr.indexOf(key) === i;
    });

    this.touched = [];
    return touched;
  }
}

export default Store;
