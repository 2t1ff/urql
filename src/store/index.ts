import { Entity, Link, Links, Records } from '../types';
import { assignObjectToMap, objectOfMap } from './utils';

export interface Store {
  records: Records;
  links: Links;
}

export interface SerializedStore {
  records: { [link: string]: Entity };
  links: { [link: string]: Link };
}

/** Creates a new Store with an optional initial, serialized state */
export const create = (initial?: SerializedStore): Store => {
  const records: Records = new Map();
  const links: Links = new Map();

  if (initial !== undefined) {
    assignObjectToMap(records, initial.records);
    assignObjectToMap(links, initial.links);
  }

  return { records, links };
};

/** Serializes a given Store to a plain JSON structure */
export const serialize = (store: Store): SerializedStore => {
  const records = objectOfMap(store.records);
  const links = objectOfMap(store.links);
  return { records, links };
};

export const find = (store: Store, key: string): Entity | null => {
  const entity = store.records.get(key);
  return entity !== undefined ? entity : null;
};

export const findOrCreate = (store: Store, key: string): Entity => {
  const entity = find(store, key);
  if (entity !== null) {
    return entity;
  }

  const record: Entity = Object.create(null);
  store.records.set(key, record);
  return record;
};

export const setLink = (store: Store, key: string, link: Link): void => {
  store.links.set(key, link);
};

export const deleteLink = (store: Store, key: string): void => {
  store.links.delete(key);
};

export const readLink = (store: Store, key: string): void | Link =>
  store.links.get(key);
