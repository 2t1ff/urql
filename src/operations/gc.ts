import { joinKeys } from '../helpers';
import { find, readLink, remove, removeLink, Store } from '../store';
import { Link } from '../types';

interface Context {
  store: Store;
  visitedLinks: Set<string>;
  visitedRecords: Set<string>;
}

export const gc = (store: Store) => {
  const visitedLinks = new Set<string>();
  const visitedRecords = new Set<string>();
  const ctx = { store, visitedLinks, visitedRecords };

  walkEntity(ctx, 'query');

  store.records.forEach((_entity, key) => {
    if (!visitedRecords.has(key)) {
      remove(store, key);
    }
  });

  store.links.forEach((_link, key) => {
    if (!visitedLinks.has(key)) {
      removeLink(store, key);
    }
  });
};

const walkEntity = (ctx: Context, key: string) => {
  const { store, visitedRecords, visitedLinks } = ctx;
  const entity = find(store, key);

  if (entity !== null && !visitedRecords.has(key)) {
    visitedRecords.add(key);

    for (const fieldKey in entity) {
      const value = entity[key];
      if (value === null) {
        const childFieldKey = joinKeys(key, fieldKey);
        const link = readLink(store, childFieldKey);
        if (link !== undefined && !visitedLinks.has(childFieldKey)) {
          visitedLinks.add(childFieldKey);
          walkLink(ctx, link);
        }
      }
    }
  }
};

const walkLink = (ctx: Context, link: Link) => {
  if (Array.isArray(link)) {
    link.forEach(childLink => walkLink(ctx, childLink));
  } else if (link !== null) {
    walkEntity(ctx, link);
  }
};
