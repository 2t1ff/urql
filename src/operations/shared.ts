import { FieldNode } from 'graphql';

import {
  getFragments,
  getMainOperation,
  getName,
  getNormalizedVars,
  getSelectionSet,
  isFieldNode,
  isInlineFragment,
  SelectionSet,
  shouldInclude,
} from '../ast';

import { Store } from '../store';
import { Context, Request } from './types';

export const makeContext = (store: Store, request: Request): void | Context => {
  const { query, variables } = request;
  const operation = getMainOperation(query);
  if (operation === undefined) {
    return;
  }

  const dependencies = [];
  const fragments = getFragments(query);
  const vars = getNormalizedVars(operation, variables);
  const isComplete = true;

  return { dependencies, isComplete, operation, fragments, vars, store };
};

export const forEachFieldNode = (
  ctx: Context,
  select: SelectionSet,
  cb: (node: FieldNode) => void
) => {
  const { vars, fragments } = ctx;

  select.forEach(node => {
    if (!shouldInclude(node, vars)) {
      // Directives instruct this node to be skipped
      return;
    } else if (!isFieldNode(node)) {
      // A fragment is either referred to by FragmentSpread or inline
      const def = isInlineFragment(node) ? node : fragments[getName(node)];

      if (def !== undefined) {
        const fragmentSelect = getSelectionSet(def);
        // TODO: Check for getTypeCondition(def) to match
        // Recursively process the fragments' selection sets
        forEachFieldNode(ctx, fragmentSelect, cb);
      }
    } else {
      cb(node);
    }
  });
};
