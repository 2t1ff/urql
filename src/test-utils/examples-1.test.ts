import gql from 'graphql-tag';
import { query, write } from '../operations';
import { create, serialize } from '../store';

const Todos = gql`
  query {
    __typename
    todos {
      __typename
      id
      text
      complete
    }
  }
`;

const ToggleTodo = gql`
  mutation($id: ID!) {
    __typename
    toggleTodo(id: $id) {
      __typename
      id
      text
      complete
    }
  }
`;

it('passes the "getting-started" example', () => {
  const store = create();
  const todosData = {
    __typename: 'Query',
    todos: [
      { id: '0', text: 'Go to the shops', complete: false, __typename: 'Todo' },
      { id: '1', text: 'Pick up the kids', complete: true, __typename: 'Todo' },
      { id: '2', text: 'Install urql', complete: false, __typename: 'Todo' },
    ],
  };

  const writeRes = write(store, { query: Todos }, todosData);

  expect(writeRes.dependencies).toEqual([
    'Query.todos',
    'Todo:0',
    'Todo:1',
    'Todo:2',
  ]);

  expect(serialize(store)).toMatchSnapshot();

  let queryRes = query(store, { query: Todos });

  expect(queryRes.data).toEqual(todosData);
  expect(queryRes.dependencies).toEqual(writeRes.dependencies);
  expect(queryRes.isComplete).toBe(true);

  const mutatedTodo = {
    ...todosData.todos[2],
    complete: true,
  };

  const mutationRes = write(
    store,
    { query: ToggleTodo, variables: { id: '2' } },
    {
      __typename: 'Mutation',
      toggleTodo: mutatedTodo,
    }
  );

  expect(mutationRes.dependencies).toEqual(['Todo:2']);
  expect(serialize(store)).toMatchSnapshot();

  queryRes = query(store, { query: Todos });

  expect(queryRes.isComplete).toBe(true);
  expect(queryRes.data).toEqual({
    ...todosData,
    todos: [...todosData.todos.slice(0, 2), mutatedTodo],
  });
});
