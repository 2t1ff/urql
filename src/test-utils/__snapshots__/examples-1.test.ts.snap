// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`passes the "getting-started" example 1`] = `
Object {
  "links": Object {
    "query.todos": Array [
      "Todo:0",
      "Todo:1",
      "Todo:2",
    ],
  },
  "records": Object {
    "Todo:0": Object {
      "__typename": "Todo",
      "complete": false,
      "id": "0",
      "text": "Go to the shops",
    },
    "Todo:1": Object {
      "__typename": "Todo",
      "complete": true,
      "id": "1",
      "text": "Pick up the kids",
    },
    "Todo:2": Object {
      "__typename": "Todo",
      "complete": false,
      "id": "2",
      "text": "Install urql",
    },
    "query": Object {
      "todos": null,
    },
  },
}
`;

exports[`passes the "getting-started" example 2`] = `
Object {
  "links": Object {
    "query.todos": Array [
      "Todo:0",
      "Todo:1",
      "Todo:2",
    ],
  },
  "records": Object {
    "Todo:0": Object {
      "__typename": "Todo",
      "complete": false,
      "id": "0",
      "text": "Go to the shops",
    },
    "Todo:1": Object {
      "__typename": "Todo",
      "complete": true,
      "id": "1",
      "text": "Pick up the kids",
    },
    "Todo:2": Object {
      "__typename": "Todo",
      "complete": true,
      "id": "2",
      "text": "Install urql",
    },
    "query": Object {
      "todos": null,
    },
  },
}
`;
